Присоединение БД

SAWINBEAST opened the pull request.

SAWINBEAST committed:
Проверка

SAWINBEAST committed:
Тест

SAWINBEAST committed:
Поиск ошибки подключения

Commit:
LHT-3 Добавлен файл Базы Данных

Commit:
LHT-3 удалены лишние файлы

SAWINBEAST committed:
Заглушка

SAWINBEAST committed:
Тест

SAWINBEAST committed:
Merge branch 'feature/LHT-3-save-habit' of https://github.com/SAWINBEAST/Life-Habit-Tracker into feature/LHT-3-save-habit

SAWINBEAST committed:
Аналог паттерна Репозиторий

Промежуточный вариант, требующий доработок.

SAWINBEAST committed:
Добавление привычки в БД

Старые данные из БД стираются.

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/Expressions/DaysTableExpression.cs:
```diff
@@ -4,7 +4,7 @@
 using System.Threading.Tasks;
 
-namespace LifeHabitTracker.DataAccessLayer.Entities
+namespace LifeHabitTracker.DataAccessLayer.Entities.Expressions
```
Лучше папку переименовать в SqlFunctions или что-то подобное

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/Expressions/HabitsTableExpression.cs:
```diff
@@ -4,16 +4,16 @@
     /// <summary>
     /// Хранитель шаблонов SQL-выражений для взаимодействия с таблицей habits
     /// </summary>
     static class HabitsTableExpression
```
Во-первых, классу нужно задать модификатор доступа (тут скорее всего internal).
Во-вторых, лучше переименовать в HabitsSqlFunctions.

SAWINBEAST replied:
Частично исправлено. Не согласен по модификатору доступа

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/Expressions/DaysTableExpression.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <summary>
+    /// Хранитель шаблонов SQL-выражений для взаимодействия с таблицей days
+    /// </summary>
+    static class DaysTableExpression
```
Во-первых, классу нужно задать модификатор доступа (тут скорее всего internal).
Во-вторых, лучше переименовать в DaysSqlFunctions.

SAWINBEAST replied:
Частично исправлено. Не согласен по модификатору доступа

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/Expressions/TimesTableExpression.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <summary>
+    /// Хранитель шаблонов SQL-выражений для взаимодействия с таблицей times
+    /// </summary>
+    static class TimesTableExpression
```
Во-первых, классу нужно задать модификатор доступа (тут скорее всего internal).
Во-вторых, лучше переименовать в TimesSqlFunctions.

SAWINBEAST replied:
Частично исправлено. Не согласен по модификатору доступа

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/PreparedData/PreparedDaysTableData.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// Класс подготовленной информации для таблицы days
+    /// </summary>
+    public class PreparedDaysTableData
```
Просто DbDays

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/PreparedData/PreparedHabitsTableData.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <summary>
+    /// Класс подготовленной информации для таблицы habits
+    /// </summary>
+    public class PreparedHabitsTableData
```
Просто DbHabits

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/PreparedData/PreparedTimesTableData.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <summary>
+    /// Класс подготовленной информации для таблицы times
+    /// </summary>
+    public class PreparedTimesTableData
```
Просто DbTimes

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// Объект выполнения записи данных в БД, в таблицу days
+        /// </summary>
+        private readonly IHabitsTableRepository _habitsManager;
```
Наименование поле всегда делай похожим на его тип - _habitsRepository.

SAWINBEAST replied:
Исправленo

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// Объект выполнения записи данных в БД, в таблицу habits
+        /// </summary>
+        private readonly IDaysTableRepository _daysManager;
```
Наименование поле всегда делай похожим на его тип - _daysRepository.

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// Объект выполнения записи данных в БД, в таблицу times
+        /// </summary>
+        private readonly ITimesTableRepository _timesManager;
```
Наименование поле всегда делай похожим на его тип - _timesRepository.

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        /// </summary>
+        private readonly ITimesTableRepository _timesManager;
+
+        public InsertHabitService(IHabitsTableRepository habitsManager, IDaysTableRepository daysManager, ITimesTableRepository timesManager)
```
Тут также не забудь переименовать.

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+                var transaction = connection.BeginTransaction();
+
+                //TODO:Оптимизировать систему проверки и не использольвать временную переменную.
+                var insertResult = new bool();
```
Неправильно. Вернись в тему про то, что такое значимые и ссылочные типы.

SAWINBEAST replied:
Исправлено

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            string connectionString = $"Data Source={dbName}";
+
+            await using (var connection = new SqliteConnection(connectionString))
```
Здесь await не нужен. Зачем?
Также в C# уже относительно давно есть возможность писать просто `using var connection = new SqliteConnection(connectionString);` и без фигурных скобок.

SAWINBEAST replied:
Исправлено

SAWINBEAST replied:
но со скобками как-то читабельнее. сразу выделяется блок

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        ///<inheritdoc/>
+        public async Task<bool> InsertHabitAsync(PreparedHabitsTableData preparedHabits, PreparedDaysTableData preparedDays, PreparedTimesTableData preparedTimes, string dbName)
+        {
+            string connectionString = $"Data Source={dbName}";
```
`var connectionString = $"Data Source={dbName}";`

SAWINBEAST replied:
ой...
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+        ///<inheritdoc/>
+        public async Task<bool> InsertHabitAsync(PreparedHabitsTableData preparedHabits, PreparedDaysTableData preparedDays, PreparedTimesTableData preparedTimes, string dbName)
+        {
+            string connectionString = $"Data Source={dbName}";
```
Полная строка соединения должна быть в appsettings. Зачем эти дополнительные манипуляции тут?

SAWINBEAST replied:
Потому, что так оптимальнее. А что, если я захочу подключаться к другой БД, например к MSSQL. Там строка подключения выглядит вообще по-другому. Да и в туториале по передаче строки подключения БД так было. 

Исправил.

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+                    var resultOfDaysInsert = await _daysManager.InsertIntoDaysTableAsync(preparedDays, habitId, connection, transaction);
+                    var resultOfTimesInsert = await _timesManager.InsertIntoTimesTableAsync(preparedTimes, habitId, connection, transaction);
+
+                    if (resultOfDaysInsert == true && resultOfTimesInsert == true)
```
Можно проще; `insertResult = resultOfDaysInsert && resultOfTimesInsert;`. Более красивый и удобный код получается.

SAWINBEAST replied:
Спасибо, исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/InsertHabitService.cs:
```diff
@@ -0,0 +1,68 @@
+                    else insertResult = false;
+
+                }
+                else if (habitId != null) insertResult = true;
```
else if или просто else на одной строке, а кусок кода входящего под это условие всегда на другой

SAWINBEAST replied:
исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/DaysTableRepository.cs:
```diff
@@ -0,0 +1,48 @@
+{
+
+    /// <inheritdoc cref="IDaysTableRepository"/>.
+    public class DaysTableRepository : IDaysTableRepository
```
Просто DaysRepository.

SAWINBEAST replied:
исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/DaysTableRepository.cs:
```diff
@@ -0,0 +1,48 @@
+        public async Task<bool> InsertIntoDaysTableAsync(PreparedDaysTableData daysTableData, long habitId, SqliteConnection connection, SqliteTransaction transaction)
+        {
+
+            var commandDaysTable = new SqliteCommand(DaysTableExpression.InsertDays, connection);
```
using

SAWINBEAST replied:
Исправил. Не понимаю, зачем тут явно прописывать освобождение ресурсов, если по окончанию работы метода InsertIntoDaysTableAsync переменная commandDaysTable у нас и так сотрётся.

SAWINBEAST replied:
Вопрос изучен и усвоен

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/DaysTableRepository.cs:
```diff
@@ -0,0 +1,48 @@
+            commandDaysTable.Parameters.Add(saturdayParam);
+            commandDaysTable.Parameters.Add(sundayParam);
+
+            if (commandDaysTable.ExecuteNonQuery() != null)
```
Посмотри, что возвращает ExecuteNonQuery и поймёшь, что не совсем правильно делаешь проверку. Это первое.
Второе. Код можно упростить.
Третье. Вот тут должен быть вызов асинхронного метода, а не синхронного. У тебя метод как асинхронный, а ты нигде ничего не ждёшь.

SAWINBEAST replied:
> Второе. Код можно упростить.

Ты намекаешь на то, что можно не писать много раз `commandDaysTable.Parameters` , а только к одной записи прописать много раз `.Add(param)`. Я тоже хотел так сделать изначально, но так нельзя. Компилятор требует так. Изучил этот момент в StackOverflow, там делают так же.

SAWINBEAST replied:
> Третье. Вот тут должен быть вызов асинхронного метода, а не синхронного. У тебя метод как асинхронный, а ты нигде ничего не ждёшь.

Я внимательно изучил работу с асинхронностью (по крайней мере, мне так кажется). Говорится, что если у нас есть дерево из асинхронных методов, то await должен стоять только на верхнем уровне, так как остальные (нижние) асинхронные методы возвращают Таску.
В моем случае await стоит в методе `InsertHabitAsync`. И в моментах `await _habitsRepository.InsertIntoHabitsTableAsync(preparedHabits, connection, transaction);` (и ему подобных) у меня как раз стоит ожидание. Ожидать commandDaysTable.ExecuteNonQuery() не имеет смысла.
Я правильно понял ?

SAWINBEAST replied:
Исправил.
Я и не знал, что существуют методы ExecuteAsync. 
Metanit в части ADO.Net и работе с SQLITE мне об этом не сказал. Как я должен был об этом догадаться ? Спасибо intallisense, что случайно предложил мне написать ExecuteNonQueryAsync.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/HabitsTableRepository.cs:
```diff
@@ -0,0 +1,35 @@
+{
+
+    /// <inheritdoc cref="IHabitsTableRepository"/>.
+    public class HabitsTableRepository : IHabitsTableRepository
```
Просто HabitsRepository.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/HabitsTableRepository.cs:
```diff
@@ -0,0 +1,35 @@
+        public async Task<long> InsertIntoHabitsTableAsync(PreparedHabitsTableData habitsTableData, SqliteConnection connection, SqliteTransaction transaction)
+        {
+
+            var commandHabitTable = new SqliteCommand(HabitsTableExpression.InsertAllFields, connection);
```
using

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/HabitsTableRepository.cs:
```diff
@@ -0,0 +1,35 @@
+            commandHabitTable.Parameters.Add(isGoodParam);
+
+
+            return (long)commandHabitTable.ExecuteScalar();
```
Посмотри, что возвращает ExecuteNonQuery и поймёшь, что не совсем правильно делаешь проверку. Это первое.
Второе. Вот тут должен быть вызов асинхронного метода, а не синхронного. У тебя метод задан как асинхронный, а ты нигде ничего не ждёшь.

SAWINBEAST replied:
Исправил, спасибо

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/TimesTableRepository.cs:
```diff
@@ -0,0 +1,38 @@
+{
+
+    /// <inheritdoc cref="ITimesTableRepository"/>.
+    public class TimesTableRepository : ITimesTableRepository
```
Просто TimesRepository.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/TimesTableRepository.cs:
```diff
@@ -0,0 +1,38 @@
+
+            foreach (var time in timesTableData.Times)
+            {
+                var commandTimeTable = new SqliteCommand(TimesTableExpression.InsertTime, connection);
```
using

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/TimesTableRepository.cs:
```diff
@@ -0,0 +1,38 @@
+                commandTimeTable.Parameters.Add(idTimeParam);
+                commandTimeTable.Parameters.Add(timeParam);
+
+                numberOfRecorders = numberOfRecorders + commandTimeTable.ExecuteNonQuery();
```
Вот тут должен быть вызов асинхронного метода, а не синхронного. У тебя метод задан как асинхронный, а ты нигде ничего не ждёшь.

SAWINBEAST replied:
Исправил, спасибо.

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/TimesTableRepository.cs:
```diff
@@ -0,0 +1,38 @@
+                numberOfRecorders = numberOfRecorders + commandTimeTable.ExecuteNonQuery();
+            }
+
+            if (numberOfRecorders == timesTableData.Times.Count) return true;
```
Красивее так: `return numberOfRecorders == timesTableData.Times.Count;`.

SAWINBEAST replied:
Благодарю, исправил.

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Interfaces/IInsertHabitService.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <param name="preparedHabits">Подготовленные основные данные о привычке</param>
+        /// <param name="preparedDays">Подготовленные данные о днях напоминания</param>
+        /// <param name="preparedTimes">Подготовленные данные о времени напоминания</param>
+        /// <param name="dbName">Строка подключения к БД</param>
```
Никаких строк подключения извне уровня данных. Эту строку подключения нужно уже в конструкторах нужных классов уровня данных получать.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Interfaces/IInsertHabitService.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// Интерфейс добавления всех данных о привычке
+    /// </summary>
+    public interface IInsertHabitService
```
В целом адекватное решение для общих задач выносить взаимодействие через общий интерфейс такой, который заточен под конкретный запрос. Одобряю.

SAWINBEAST replied:
Спасибо

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Interfaces/Repositories/IDaysTableRepository.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <summary>
+    /// Интерфейс взаимодействия с таблицей days
+    /// </summary>
+    public interface IDaysTableRepository
```
Просто IDaysRepository.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Interfaces/Repositories/IHabitsTableRepository.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// Интерфейс взаимодействия с таблицей habits
+    /// </summary>
+    public interface IHabitsTableRepository
```
Просто IHabitsRepository.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Interfaces/Repositories/ITimesTableRepository.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Интерфейс взаимодействия с таблицей times
+    /// </summary>
+    public interface ITimesTableRepository
```
Просто ITimesRepository.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/IP_Log_Data.js:
```diff
@@ -0,0 +1,200 @@
+data='\
+28-7-2023,10:23:11,192.168.1.126,\n\
```
Что за странный файл? Удали

SAWINBEAST replied:
Странно , не могу его найти у себя.
Файл JavaScript, к тому же ...

SAWINBEAST replied:
Удалил

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Entities/DataBaseConnect.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace LifeHabitTracker.BusinessLogicLayer.Entities
+{
+    public class DataBaseConnect
```
Класс вообще никаким боком не относится к уровню бизнес-логики. В принципе дефолтная сущность. Для такого можно сделать отдельный проект Entities и туда закинуть этот класс. Дёргать его, где надо.

SAWINBEAST replied:
Мне кажется, лучше закинуть в уровень DataAccess  в папку Entities, ведь теперь у меня передаётся строка подключения к БД сразу в уровень данных, а не в бизнес-логику. Что думаешь ?

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/Habit.cs:
```diff
@@ -25,7 +25,7 @@ public class Habit
         /// Даты выполнения привычки
         /// </summary>
-        public ReminderDate Date { get; set; }
+        public ReminderDate Date { get; set; } = null;
```
Лишнее

SAWINBEAST replied:
исправил

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// <summary>
+        /// Информация о подключаемой БД
+        /// </summary>
+        public static DataBaseConnect _dBConfig;
```
Это точно не на уровне бизнес-логики.

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// </summary>
+        private readonly Dictionary<string, string> _daysTypesOfNames = new Dictionary<string, string>()
+        {
+            {"пн", "monday"},
```
У тебя же есть класс с константами. Сделай ещё такой же с английской версией (его тоже лучше в отдельный проект Entities, так как эти константы тебе нужны и на уровне данных).

SAWINBEAST replied:
Не понял. 
Ты советуешь мне сделать константы английских названий дней недели ?
Не могу увидеть надобность. Если в конце концов объединение будет по такому же смыслу, то у меня просто получится больше кода и менее читабельный словарь `_daysTypesOfNames`

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// <summary>
+        /// Объект подготовленных данных для таблицы habits
+        /// </summary>
+        private PreparedHabitsTableData _habitsTableData = new PreparedHabitsTableData();
```
Не нужно на уровне члена-класса. Это никак не связано с описанием самого класса.

SAWINBEAST replied:
Но это же всё равно приватное поле. Оно будет использовано только внутри класса, для внутренних нужд.

И к тому же, учитывая, что у меня подготовка данных под таблицы БД выведена в отдельный метод `PrepareDataForTable(habit, chatId);` , а потом ещё эти изменённые данные отправляются на запись в `InsertHabitAsync` из метода `AddHabitAsync` , то вполне логично использовать эти сущности (тип которых уже называется DbHabits) на уровне класса. 

Я правильно понимаю ?

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// <summary>
+        /// Объект подготовленных данных для таблицы days
+        /// </summary>
+        private PreparedDaysTableData _daysTableData = new PreparedDaysTableData();
```
Не нужно на уровне члена-класса. Это никак не связано с описанием самого класса.

SAWINBEAST replied:
Но это же всё равно приватное поле. Оно будет использовано только внутри класса, для внутренних нужд.

И к тому же, учитывая, что у меня подготовка данных под таблицы БД выведена в отдельный метод `PrepareDataForTable(habit, chatId);` , а потом ещё эти изменённые данные отправляются на запись в `InsertHabitAsync` из метода `AddHabitAsync` , то вполне логично использовать эти сущности (тип которых уже называется DbHabits) на уровне класса. 

Я правильно понимаю ?

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// <summary>
+        /// Объект подготовленных данных для таблицы times
+        /// </summary>
+        private PreparedTimesTableData _timesTableData = new PreparedTimesTableData();
```
Не нужно на уровне члена-класса. Это никак не связано с описанием самого класса.

SAWINBEAST replied:
Но это же всё равно приватное поле. Оно будет использовано только внутри класса, для внутренних нужд.

И к тому же, учитывая, что у меня подготовка данных под таблицы БД выведена в отдельный метод `PrepareDataForTable(habit, chatId);` , а потом ещё эти изменённые данные отправляются на запись в `InsertHabitAsync` из метода `AddHabitAsync` , то вполне логично использовать эти сущности (тип которых уже называется DbHabits) на уровне класса. 

Я правильно понимаю ?

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+        /// </summary>
+        /// <param name="habit"> Привычка уровня бизнесЛогики </param>
+        /// <param name="chatId">ID чата, из которого добавляют привычку</param>
+        private void PrepareDataForTable(Habit habit, long chatId)
```
Под каждую сущность можно написать отдельный метод по формированию сущности на сохранение в базу данных. Соответственно каждый метод должен возвращать сущность.

SAWINBEAST replied:
Не исправил.
Жду координации по вопросу с полем класса `private PreparedTimesTableData _timesTableData = new PreparedTimesTableData();` (и ему подобных)

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -1,33 +1,144 @@
+            _habitsTableData.Description = habit.Description;
+            _habitsTableData.ChatId = chatId;
+
+            var type = () => (habit.Type == "хорошая") ? 1 : 0;
```
Зачем?
`_habitsTableData.IsGood = habit.Type == "хорошая";`
И насколько я помню у тебя под "хорошая" или "плохая" тоже есть класс с этими константами. Используй их.

SAWINBEAST replied:
Исправил, спасибо

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.sqbpro:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?><sqlb_project><db path="C:/Users/Саша/Desktop/LifeHabitTracker.db" readonly="0" foreign_keys="1" case_sensitive_like="0" temp_store="0" wal_autocheckpoint="1000" synchronous="2"/><attached/><window><main_tabs open="structure browser pragmas query" current="1"/></window><tab_structure><column_width id="0" width="300"/><column_width id="1" width="0"/><column_wi ...
```
Что-то лишнее

SAWINBEAST replied:
Это часть БД. если я правильно понимаю. Файл, без которого запись не будет осуществляться

MAXon28 commented on LifeHabitTracker/Network_Meter_Data.js:
```diff
@@ -0,0 +1,345 @@
+data='\
```
Что-то лишнее

SAWINBEAST replied:
Удалил

SAWINBEAST committed:
Общий Рефакторинг + Внешние сервисы + Библиотека Сущностей

MAXon28 committed:
LHT-3 Улучшение кода

1. Улучшил обработку данных по дням.
2. Убрал лишнее.
3. Настроил конфигурацию.
4. Удалил по возможности в файлах непонятные энтеры.

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Entities/SqlFunctions/HabitsSqlFunctions.cs:
```diff
@@ -0,0 +1,17 @@
+        public const string InsertAllFields = "INSERT INTO habits (name, desc, chat_id, is_good) " +
+                                                  "VALUES (@name, @desc, @chatid, @isgood); " +
+                                                  "SELECT last_insert_rowid();";
+
```
Удали лишнюю строку

SAWINBEAST replied:
Исправил

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/DaysRepository.cs:
```diff
@@ -0,0 +1,44 @@
+            commandDaysTable.Parameters.Add(saturdayParam);
+            commandDaysTable.Parameters.Add(sundayParam);
+
+            if (await commandDaysTable.ExecuteNonQueryAsync() != 0)
```
Уже говорил, как можно проще и лаконичнее писать - вспоминай.

SAWINBEAST replied:
Исправил.
Забыл, Благодарю

MAXon28 commented on LifeHabitTracker/DataAccessLayer/Impls/Repositories/TimesRepository.cs:
```diff
@@ -0,0 +1,33 @@
+                commandTimeTable.Parameters.Add(idTimeParam);
+                commandTimeTable.Parameters.Add(timeParam);
+
+                numberOfRecorders +=  await commandTimeTable.ExecuteNonQueryAsync();
```
Лишний пробел

SAWINBEAST replied:
Глаз алмаз)
Исправил.
Прививаю принцип

SAWINBEAST committed:
Подключение БД / Окончательный вариант
